dpctl/memory/_sycl_usm_array_interface_utils.pxi

```diff
@@ -0,0 +1,187 @@
+    if (_valid_usm_ptr_and_context(ptr, ctx)):
+        return DPCTLQueue_Copy(q.get_queue_ref())
+    else:
+        return NULL
+
+
+cdef DPCTLSyclQueueRef _queue_ref_copy_from_USMRef_and_SyclContext(
+    DPCTLSyclUSMRef ptr, SyclContext ctx):
+    """ Obtain device from pointer and sycl context, use
+        context and device to create a queue from which this memory
+        can be accessible.
+    """
+    cdef SyclDevice dev = _Memory.get_pointer_device(ptr, ctx)
+    cdef DPCTLSyclContextRef CRef = ctx.get_context_ref()
+    cdef DPCTLSyclDeviceRef DRef = dev.get_device_ref()
+    return DPCTLQueue_Create(CRef, DRef, NULL, 0)
+
+
+cdef DPCTLSyclQueueRef get_queue_ref_from_ptr_and_syclobj(
+    DPCTLSyclUSMRef ptr, object syclobj):
+    """ Constructs queue from pointer and syclobject from
+        __sycl_usm_array_interface__
+    """
+    cdef DPCTLSyclQueueRef QRef = NULL
+    cdef SyclContext ctx
+    if type(syclobj) is SyclQueue:
+        return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> syclobj)
+    elif type(syclobj) is SyclContext:
+        ctx = <SyclContext>syclobj
+        return _queue_ref_copy_from_USMRef_and_SyclContext(ptr, ctx)
+    elif type(syclobj) is str:
+        q = SyclQueue(syclobj)
+        return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> q)
+    elif pycapsule.PyCapsule_IsValid(syclobj, "SyclQueueRef"):
+        q = SyclQueue(syclobj)
+        return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> q)
+    elif pycapsule.PyCapsule_IsValid(syclobj, "SyclContextRef"):
+        ctx = <SyclContext>SyclContext(syclobj)
+        return _queue_ref_copy_from_USMRef_and_SyclContext(ptr, ctx)
+    elif hasattr(syclobj, '_get_capsule'):
+        cap = syclobj._get_capsule()
+        if pycapsule.PyCapsule_IsValid(cap, "SyclQueueRef"):
+            q = SyclQueue(cap)
+            return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> q)
+        elif pycapsule.PyCapsule_IsValid(cap, "SyclContexRef"):
+            ctx = <SyclContext>SyclContext(cap)
+            return _queue_ref_copy_from_USMRef_and_SyclContext(ptr, ctx)
+        else:
+            return QRef
+    else:
+        return QRef
+
+
+cdef object _pointers_from_shape_and_stride(
+    int nd, object ary_shape, Py_ssize_t itemsize, Py_ssize_t ary_offset,
+    object ary_strides):
+    """
+    Internal utility: for given array data about shape/layout/element
+    compute left-most displacement when enumerating all elements of the array
+    and the number of bytes of memory between the left-most and right-most
+    displacements.
+
+    Returns: tuple(min_disp, nbytes)
+    """
+    if (nd > 0):
+        if (ary_strides is None):
+            nelems = 1
+            for si in ary_shape:
+                sh_i = int(si)
+                if (sh_i <= 0):
+                    raise ValueError("Array shape elements need to be positive")
+                nelems = nelems * sh_i
+            return (ary_offset, nelems * itemsize)
```

Comment:
What is the expected value for `ary_offset` in the default case when strides are based on lexicographic traversal? Is it always going to be 0? If not zero, will be not have a potential out of bound access?

Reply:
Most common case of the offset is 0, most common non-zero offset is when negative strides are involved:

```
In [1]: import numpy as np

In [2]: X = np.empty((4,5))

In [5]: X[::-1, ::-1].ctypes.data
Out[5]: 93992765649928

In [6]: X[::-1, ::-1].ctypes.data - X.ctypes.data
Out[6]: 152
```

This corresponds to (4*5-1)*8, where 8 is the itemsize of the array element.